vir/src/callable_idents.rs

```diff
@@ -0,0 +1,140 @@
+use crate::{ExprGen, PredicateAppGen, Type, PredicateAppGenData, StmtGenData, MethodCallGenData, VirCtxt};
+use private::*;
+
+#[derive(Debug, Clone, Copy)]
+pub struct FunctionIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> FunctionIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct MethodIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> MethodIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PredicateIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> PredicateIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+/// Module to prevent anyone else from implementing Arity
+mod private {
+    use super::*;
+    pub trait Arity {}
+    impl<const N: usize> Arity for KnownArity<'_, N> {}
+    impl Arity for UnknownArity<'_> {}
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct KnownArity<'vir, const N: usize>([Type<'vir>; N]);
```

Comment:
The array itself is not arena-allocated, should this store a slice instead?

Reply:
Since the array is fixed width it's always copy so doesn't need to be arena allocd. Also, I don't expect to have functions of known arity of more than 2 or 3 arguments and so it would be a waste to store this in the arena (imagine that it's essentially a tuple).